lib/grpc_generator.dart

```diff
@@ -0,0 +1,252 @@
+// BSD-style license that can be found in the LICENSE file.
+
+part of protoc;
+
+class GrpcServiceGenerator {
+  final ServiceDescriptorProto _descriptor;
+
+  /// The generator of the .pb.dart file that will contain this service.
+  final FileGenerator fileGen;
+
+  /// The message types needed directly by this service.
+  ///
+  /// The key is the fully qualified name.
+  /// Populated by [resolve].
+  final _deps = <String, MessageGenerator>{};
+
+  /// Maps each undefined type to a string describing its location.
+  ///
+  /// Populated by [resolve].
+  final _undefinedDeps = <String, String>{};
+
+  /// Fully-qualified gRPC service name.
+  String _fullServiceName;
+
+  /// Dart class name for client stub.
+  String _clientClassname;
+
+  /// Dart class name for server stub.
+  String _serviceClassname;
+
+  /// List of gRPC methods.
+  final _methods = <_GrpcMethod>[];
+
+  GrpcServiceGenerator(this._descriptor, this.fileGen) {
+    final name = _descriptor.name;
+    final package = fileGen.package;
+
+    if (package != null && package.isNotEmpty) {
+      _fullServiceName = '$package.$name';
+    } else {
+      _fullServiceName = name;
+    }
+
+    // avoid: ClientClient
+    _clientClassname = name.endsWith('Client') ? name : name + 'Client';
+    // avoid: ServiceServiceBase
+    _serviceClassname =
+        name.endsWith('Service') ? name + 'Base' : name + 'ServiceBase';
+  }
+
+  /// Finds all message types used by this service.
+  ///
+  /// Puts the types found in [_deps]. If a type name can't be resolved, puts it
+  /// in [_undefinedDeps].
+  /// Precondition: messages have been registered and resolved.
+  void resolve(GenerationContext ctx) {
+    for (var method in _descriptor.method) {
+      _methods.add(new _GrpcMethod(this, method)..resolve(ctx));
+    }
+  }
+
+  /// Adds a dependency on the given message type.
+  ///
+  /// If the type name can't be resolved, adds it to [_undefinedDeps].
+  void _addDependency(GenerationContext ctx, String fqname, String location) {
+    if (_deps.containsKey(fqname)) return; // Already added.
+
+    MessageGenerator mg = ctx.getFieldType(fqname);
+    if (mg == null) {
+      _undefinedDeps[fqname] = location;
+      return;
+    }
+    mg.checkResolved();
+    _deps[mg.fqname] = mg;
+  }
+
+  /// Adds dependencies of [generate] to [imports].
+  ///
+  /// For each .pb.dart file that the generated code needs to import,
+  /// add its generator.
+  void addImportsTo(Set<FileGenerator> imports) {
+    for (var mg in _deps.values) {
+      imports.add(mg.fileGen);
+    }
+  }
+
+  /// Returns the Dart class name to use for a message type.
+  ///
+  /// Throws an exception if it can't be resolved.
+  String _getDartClassName(String fqname) {
+    var mg = _deps[fqname];
+    if (mg == null) {
+      var location = _undefinedDeps[fqname];
+      throw 'FAILURE: Unknown type reference (${fqname}) for ${location}';
```

Comment:
At some point, it might be nice to start maintaining a list of all possible "failures" and to work on making the error text associated with them easy to grok and actionable. For now, throwing a string will probably do, but maybe add a TODO here.